monkeytype/typing.py

```diff
@@ -58,31 +88,47 @@ def __are_typed_dict_types_equal(type1: type, type2: type) -> bool:
 _TypedDictMeta.__eq__ = __are_typed_dict_types_equal
 
 
-def typed_dict_to_dict(type_dict: type) -> type:
-    """Convert a TypedDict to a Dict.
-    The keys of a TypedDict that we have constructed will always be strings.
-    However, if the TypedDict is empty, we return the key type as Any since we
-    can't justify a key type."""
-    if type_dict.__annotations__ == {}:
-        return Dict[Any, Any]
-    return Dict[str, shrink_types(type_dict.__annotations__.values())]
-
-
-def shrink_typed_dict_types(types: List[type]) -> type:
-    """Shrink the TypedDicts to a single TypedDict *only* if they are all the same.
-    Doing this manually because Union[TypedDict('Foo', {'a': int}),
-    TypedDict('Foo', {'a': int})] doesn't return a single TypedDict.
-
-    If there is any mismatch, fall back to a Union of the Dicts."""
-    assert len(types) != 0
-    first_type_dict = types[0]
-    if all(typ == first_type_dict for typ in types[1:]):
-        return first_type_dict
+def remove_requiredness_prefix(name: str) -> str:
+    if name.startswith(REQUIRED_FIELD_PREFIX):
+        prefix = REQUIRED_FIELD_PREFIX
+    elif name.startswith(NON_REQUIRED_FIELD_PREFIX):
+        prefix = NON_REQUIRED_FIELD_PREFIX
     else:
-        return Union[tuple(typed_dict_to_dict(typ) for typ in types)]
-
-
-def shrink_types(types):
+        prefix = ''
+    return name[len(prefix):]
+
+
+def shrink_typed_dict_types(typed_dicts: List[type], max_typed_dict_size: int) -> type:
+    """Shrink a list of TypedDicts into one with the required fields and the non-required fields.
+    Required fields are keys that appear in all the TypedDicts.
+    Non-required fields are those that appear in only some of the TypedDicts.
+    If the same key has a different value type, then its value is the Union of the value types."""
+    num_typed_dicts = len(typed_dicts)
+    key_value_types_dict = defaultdict(list)
+    for typed_dict in typed_dicts:
+        for key, value_type in typed_dict.__annotations__.items():
+            # Assumption: An input TypedDict to shrink_types will have all
+            # fields as required, since it was constructed directly from a
+            # dictionary object.
+            assert key.startswith(REQUIRED_FIELD_PREFIX)
+
+            key_value_types_dict[key[len(REQUIRED_FIELD_PREFIX):]].append(value_type)
+
+    required_fields = {key: shrink_types(tuple(value_types), max_typed_dict_size)
+                       for key, value_types in key_value_types_dict.items()
+                       if len(value_types) == num_typed_dicts}
+    non_required_fields = {key: shrink_types(tuple(value_types), max_typed_dict_size)
+                           for key, value_types in key_value_types_dict.items()
+                           if len(value_types) != num_typed_dicts}
+
+    if len(required_fields) + len(non_required_fields) > max_typed_dict_size:
+        value_types = (RewriteTypedDictToDict().rewrite(value_type)
```

Comment:
Also here using a recursive rewriter seems wrong. Just because the top-level dict we are shrinking is too big to be represented as a `TypedDict` doesn't imply that all nested `TypedDict` should also be converted to `Dict`.

Reply:
Yeah, I was trying to avoid sending `shrink_types` a TypedDict with optional fields because I didn't have a good idea of how to handle that case. Have fixed it and added tests :)